src/stage1/ir.cpp

```diff
@@ -30396,15 +30396,21 @@ static void buf_write_value_bytes(CodeGen *codegen, uint8_t *buf, ZigValue *val)
         case ZigTypeIdFloat:
             float_write_ieee597(val, buf, codegen->is_big_endian);
             return;
-        case ZigTypeIdPointer:
-            if (val->data.x_ptr.special == ConstPtrSpecialHardCodedAddr) {
-                BigInt bn;
-                bigint_init_unsigned(&bn, val->data.x_ptr.data.hard_coded_addr.addr);
-                bigint_write_twos_complement(&bn, buf, codegen->builtin_types.entry_usize->data.integral.bit_count, codegen->is_big_endian);
-                return;
-            } else {
-                zig_unreachable();
+        case ZigTypeIdPointer: {
+            BigInt bn;
+            switch (val->data.x_ptr.special) {
+                case ConstPtrSpecialHardCodedAddr:
+                    bigint_init_unsigned(&bn, val->data.x_ptr.data.hard_coded_addr.addr);
+                    break;
+                case ConstPtrSpecialNull:
+                case ConstPtrSpecialInvalid:
```

Comment:
```suggestion
```
`ConstPtrSpecialInvalid` is there so that zero-initialized variables have an invalid value.
https://github.com/ziglang/zig/blob/9f01598a498ecbd868cb81dc1efdb1507010c1b8/src/stage1/all_types.hpp#L238-L239
Undefined is signalled by `ConstValSpecialUndef`.

Reply:
Hmm, the following program fails if I remove this line (and it is not fixed by adding `case ConstValSpecialUndef`):

```zig
fn zero() [*c]u8 {
    var ret: [*c]u8 = undefined;
    @ptrCast(*[8]u8, &ret)[1] = 123;  // Succeeds if you remove this.
    return ret;
}

export const foo = zero();
```

I wonder if there is a bug somewhere else in the compiler where `ConstValSpecialUndef` is not getting set appropriately.

Reply:
I should have been clearer. `ConstValSpecialUndef` belongs to a different enum (notice Val instead of Ptr). See `ZigValue.special`.

Reply:
Also keep in mind that the plan is to eventually throw away the C++ code: #6378.
To fix some of the bugs you would need to rewrite quite a bit of code.

Reply:
> I should have been clearer. ConstValSpecialUndef belongs to a different enum (notice Val instead of Ptr).

I see. But it looks like `ConstPtrValue.special` is the discriminator that must be valid before we can read any other member of `ConstPtrValue.data`, no? It seems like we have to read `ConstPtrValue.data` before we can test it against `ConstValSpecialUndef`, but we can't do that without a valid discriminator. So it still seems like a bug that `val->data.x_ptr.special == ConstPtrSpecialInvalid`.

> To fix some of the bugs you would need to rewrite quite a bit of code.

Hmm, that's too bad. I guess it's throwaway work to fix bugs here if everything is going to be rewritten in Zig?